src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -53,33 +53,48 @@ public GeneratorDriver RunFullGeneration(Compilation compilation, out Compilatio
-            if (compilation == _state.Compilation && _state.FinalCompilation is object)
-            {
-                outputCompilation = _state.FinalCompilation;
-                return this;
-            }
-
             // run the actual generation
             var state = StateWithPendingEditsApplied(_state);
             var stateBuilder = PooledDictionary<ISourceGenerator, GeneratorState>.GetInstance();
+            var receivers = PooledDictionary<ISourceGenerator, ISyntaxReceiver>.GetInstance();
 
-            //PROTOTYPE: should be possible to parallelize this
             foreach (var generator in state.Generators)
             {
                 // initialize the generator if needed
                 if (!state.GeneratorStates.TryGetValue(generator, out GeneratorState generatorState))
                 {
                     generatorState = InitializeGenerator(generator, cancellationToken);
                 }
+                stateBuilder.Add(generator, generatorState);
 
+                // create the syntax receiver if requested
+                if (generatorState.Info.SyntaxReceiverCreator is object)
+                {
+                    var rx = generatorState.Info.SyntaxReceiverCreator();
+                    receivers.Add(generator, rx);
+                }
+            }
+
+            // Run a syntax walk if any of the generators requested it
+            if (receivers.Count > 0)
+            {
+                GeneratorSyntaxWalker walker = new GeneratorSyntaxWalker(receivers.Values.ToImmutableArray());
+                foreach (var syntaxTree in compilation.SyntaxTrees)
+                {
+                    walker.Visit(syntaxTree.GetRoot());
+                }
+            }
+
+            //PROTOTYPE: should be possible to parallelize this
+            foreach (var (generator, generatorState) in stateBuilder.ToImmutableDictionary())
+            {
                 try
                 {
                     // we create a new context for each run of the generator. We'll never re-use existing state, only replace anything we have
-                    var context = new SourceGeneratorContext(state.Compilation, new AnalyzerOptions(state.AdditionalTexts.NullToEmpty(), CompilerAnalyzerConfigOptionsProvider.Empty));
+                    _ = receivers.TryGetValue(generator, out var syntaxReceiverOpt);
+                    var context = new SourceGeneratorContext(state.Compilation, new AnalyzerOptions(state.AdditionalTexts.NullToEmpty(), CompilerAnalyzerConfigOptionsProvider.Empty), syntaxReceiverOpt);
                     generator.Execute(context);
-                    stateBuilder.Add(generator, generatorState.WithSources(context.AdditionalSources.ToImmutableAndFree()));
+                    stateBuilder[generator] = generatorState.WithSources(ParseAdditionalSources(context.AdditionalSources.ToImmutableAndFree()));
```

Comment:
>context.AdditionalSources [](start = 96, length = 25)

Does `AdditionalSources` need to be mutable?

Reply:
Yes the generator adds/removes source to it as part of generation. We create a new one per generator + run though, so it's safe.

---
In reply to: [395317865](https://github.com/dotnet/roslyn/pull/42507#discussion_r395317865) [](ancestors = 395317865)